sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala

```diff
@@ -658,6 +660,34 @@ object TypeCoercion {
     }
   }
 
+  /**
+   * Coerces the types of [[Concat]] children to expected ones.
+   *
+   * If `spark.sql.function.concatBinaryAsString` is false and all children types are binary,
+   * the expected types are binary. Otherwise, the expected ones are strings.
+   */
+  case class ConcatCoercion(conf: SQLConf) extends TypeCoercionRule {
+
+    private def typeCastToString(c: Concat): Concat = {
+      val newChildren = c.children.map { e =>
+        ImplicitTypeCasts.implicitCast(e, StringType).getOrElse(e)
+      }
+      c.copy(children = newChildren)
+    }
+
+    override protected def coerceTypes(plan: LogicalPlan): LogicalPlan = plan transform { case p =>
+      p transformExpressionsUp {
+        // Skip nodes if unresolved or empty children
+        case c @ Concat(children) if !c.childrenResolved || children.isEmpty => c
+
```

Comment:
Remove this line?

Reply:
Probably, we cant cuz we hit unresolved calls.

Reply:
The empty line.

Reply:
ok